extensions/azurecore/src/account-provider/azureAccountProvider.ts

```diff
@@ -159,29 +159,35 @@ export class AzureAccountProvider implements azdata.AccountProvider, vscode.Disp
-				if (this.isAuthenticationResult(authResult) && authResult.account && authResult.account.idTokenClaims) {
-					const token: Token = {
-						key: authResult.account.homeAccountId,
-						token: authResult.accessToken,
-						tokenType: authResult.tokenType,
-						expiresOn: authResult.account.idTokenClaims.exp!,
-						tenantId: tenantId,
-						resource: resource
-					};
-					try {
-						await this.msalCacheProvider.writeTokenToLocalCache(token);
-					} catch (e) {
-						Logger.error(`Could not save access token to local cache: ${e}, this might cause throttling of AAD requests.`);
-					}
-					return token;
+				if (getTenantIgnoreList().includes(tenantId)) {
+					// Tenant found in ignore list, don't fetch access token.
+					Logger.info(`Tenant ${tenantId} found in the ignore list, authentication will not be attempted. Please remove tenant from setting: '${Constants.AzureTenantConfigFilterSetting}' if you want to re-enable tenant for authentication.`);
+					throw new Error(localize('tenantIgnoredError', '{0}: Tenant found in ignore list, authentication not attempted. You can remove tenant {1} from ignore list in settings.json file: {2} if you wish to access resources from this tenant.', Constants.TenantIgnoredErrorCode, tenantId, Constants.AzureTenantConfigFilterSetting));
 				} else {
-					Logger.error(`MSAL: getToken call failed`);
-					// Throw error with MSAL-specific code/message, else throw generic error message
-					if (this.isProviderError(authResult)) {
-						throw new Error(localize('msalTokenError', `{0} occurred when acquiring token. \n{1}`, authResult.errorCode, authResult.errorMessage));
+					let authResult = await azureAuth.getTokenMsal(account.key.accountId, resource, tenantId);
+					if (this.isAuthenticationResult(authResult) && authResult.account && authResult.account.idTokenClaims) {
+						const token: Token = {
+							key: authResult.account.homeAccountId,
+							token: authResult.accessToken,
+							tokenType: authResult.tokenType,
+							expiresOn: authResult.account.idTokenClaims.exp!,
+							tenantId: tenantId,
+							resource: resource
+						};
+						try {
+							await this.msalCacheProvider.writeTokenToLocalCache(token);
+						} catch (e) {
+							Logger.error(`Could not save access token to local cache: ${e}, this might cause throttling of AAD requests.`);
+						}
+						return token;
 					} else {
-						throw new Error(localize('genericTokenError', 'Failed to get token'));
+						Logger.error(`MSAL: getToken call failed`);
+						// Throw error with MSAL-specific code/message, else throw generic error message
+						if (this.isProviderError(authResult)) {
+							throw new Error(localize('msalTokenError', `{0} occurred when acquiring token. \n{1}`, authResult.errorCode, authResult.errorMessage));
+						} else {
+							throw new Error(localize('genericTokenError', 'Failed to get token'));
```

Comment:
There isn't any other information we can add here to help the user understand why this failed? Are we usually just expected all errors to be provider errors?

Reply:
This is the case where getTokenMSAL would return 'null' - it was also encapsulating the case for ignore tenants, which shouldn't have been the case. But for other reasons, which could be many, we still need to throw generic error.

Reply:
Ok, although it feels like we should never return null if the call fails. Shouldn't there always be *some* sort of error that getTokenMSAL can throw instead with information about why it failed?

Reply:
I was thinking that too, will do it in next PR!